### ISSUE TITLE
MAC OS X: Tkinter import doesn't copy Tcl and TK

### ISSUE THREAD

**Issue by wahid** [pyinstaller/pyinstaller#1350]
On mac PyInstaller doesn't seem to copy Tcl and Tk once both are copied to the folder the error is gone.

```
module = imp.load_module(fullname, fp, filename, self._c_ext_tuple)
ImportError: dlopen(_tkinter.so, 2): Library not loaded: @loader_path/Tcl
  Referenced from: _tkinter.so
  Reason: image not found
```

**Comment by htgoebel** [comment]
Please provide more information, see https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#information-we-need

**Comment by wahid** [comment]
The error occurs when running the application.

Python: 2.7.10 
PyInstaller: 2.1
Mac OSX: Yosemite 10.10.4

You can reproduce the error by creating a script with:

```
import TKinter import *

root = Tk()
```

`Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/jphnghknbv/Desktop/python_script/build/script/out00-PYZ.pyz/Tkinter", line 39, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 409, in load_module
    module = imp.load_module(fullname, fp, filename, self._c_ext_tuple)
ImportError: dlopen(/Users/jphnghknbv/Desktop/python_script/dist/script/_tkinter.so, 2): Library not loaded: @loader_path/Tcl
  Referenced from: /Users/jphnghknbv/Desktop/python_script/dist/script/_tkinter.so
  Reason: image not found`

**Comment by htgoebel** [comment]
Please try the latest development version.

**Comment by wahid** [comment]
>Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/PyInstaller-2.1.1.dev0-py2.7.egg/PyInstaller/loader/pyi_importers.py", line 271, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/jphnghknbv/Desktop/test/build/test/out00-PYZ.pyz/Tkinter", line 39, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/PyInstaller-2.1.1.dev0-py2.7.egg/PyInstaller/loader/pyi_importers.py", line 414, in load_module
    module = imp.load_module(fullname, fp, filename, self._c_ext_tuple)
ImportError: dlopen(/Users/jphnghknbv/Desktop/test/dist/test/_tkinter.so, 2): Library not loaded: @loader_path/Tcl
  Referenced from: /Users/jphnghknbv/Desktop/test/dist/test/_tkinter.so
  Reason: no suitable image found.  Did find:
	/Users/jphnghknbv/Desktop/test/dist/test/Tcl: not a file
	/Users/jphnghknbv/Desktop/test/dist/test/Tcl: not a file

**Comment by stefsmeets** [comment]
I'm encountering the same problem. Running pyinstaller (2.1) on:

    import Tkinter
    Tkinter._test()

results in the same errors described by wahid. Is there a work-around for this?

**Comment by wahid** [comment]
What has worked for me is copying Tcl binary from "/System/Library/Frameworks/Tcl.framework" and Tk binary from "/System/Library/Frameworks/Tk.framework" to the output folder after the process completed.

**Comment by htgoebel** [comment]
Sorry, I pojnted you to the wrong version to use. Please use the current development version from the `python3` branch. See https://github.com/pyinstaller/pyinstaller/wiki#downloads for installation instructions.

**Comment by stefsmeets** [comment]
@htgoubel Using the python3 version results in the same errors.
@wahid This seems to work well, thanks!

**Comment by stefsmeets** [comment]
Just for your information, I reinstalled python via homebrew, following a suggestion elsewhere, and it magically solves this problem. I don't like this solution, but with everything runs as I would expect. I was using the standard PSF python before.

**Comment by matysek** [comment]
@stefsmeets Do you know if python from homebrew uses tcl/tk library from the OS or it compiles it's own version?

My guess is this:
* Pyinstaller does not bundle tcl/tk from the system Python, because it is considered as a system library.
* PyInstaller however modifies the path where _tkinter.so will load tcl/tk library. It modifies it to: `@loader_path/Tcl` - it means look for `Tcl` in the same directory as the final executable.
* The fix would be either:
  * bundle the system version of tcl/tk or
  * set the path where _tkinter.so is looking for Tcl to absolute path - use the system version
```
@loader_path/Tcl  ->  /System/Library/Frameworks/Tk.framework/Tk
```

**Comment by stefsmeets** [comment]
@matysek Homebrew python will use the system Tk/Tcl (version 8.5.9) that ships with OS X 10.10.

**Comment by jamb0ss** [comment]
any ideas how to fix it? I bundle my program into a single file, so I just cannot use wahid's solution... Also I've tried to install Python via homebrew - the same issue.

**Comment by wahid** [comment]
Have you tried to manually copy the files into the app file it creates?

**Comment by jamb0ss** [comment]
Where I should put this files there?

**Comment by jamb0ss** [comment]
Any ideas how to bundle Tk/Tcl-based program into a single file on OS X?

**Comment by wahid** [comment]
if it creates a *.app file I would assume you can do.

Show package contents -> Contents/MacOS or Contents/Resources and place them in one of those folders just look for the folder where there are files with the 'so' extension.

**Comment by jamb0ss** [comment]
Did it work for you? I just tried to put this files in all directories in .app file (in all at once, copy-paste) with no success, that I've mentioned before

**Comment by titan-ae** [comment]
I can confirm with latest pip pyinstaller, tkInter & Tcl do not package correctly on OS/X :-1: 

I found editing hook-_tkinter.py & pyi_rth_tkinter.py as defined here -

https://github.com/viotti/pyinstaller/commit/597af5cb01c064620b53ea1ee537e30f56fa481d

Then copying manually the following files into the dist area works -

/System/Library/Frameworks/Tcl.framework/Versions/8.5/Tcl
/System/Library/Frameworks/Tk.framework/Tk

Now it runs :+1:

**Comment by LevN0** [comment]
I would have expected that the HomeBrew method would have worked. In my case I also needed to package matplotlib, and that did not work with HomeBrew. So I used another workaround that should work,

I used the anaconda distribution. This packaged things up fine, including Tk/TCL, except that a lot of extraneous packages were being included. So I deleted most things from anaconda (via a combination of pip and just outright deleting things from site-packages) except the specific packages I wanted PyInstaller to include. Then things worked fine.

**Comment by josePhoenix** [comment]
This appears to still be an issue. I am using Python from Homebrew, but it did not solve the problem for me. About to try with Anaconda, since I can switch to that environment easily.

**Comment by scholer** [comment]
@josePhoenix I've just myself tried to get PyInstaller to work with Anaconda distribution on OS X -- no luck. The tcl library is still not properly included. (I'm thinking of trying the python.org distribution next.) Let us know if you succeed in making it work with Anaconda!

**Comment by LevN0** [comment]
The Anaconda solution as I described above still works for me, on two different machines with different OSX versions. Admittedly it is rather convoluted. At this point I have to use an old version of conda (I use v3.14.1), because newer versions seem to have other issues with PyInstaller. However, using that version Tk is packaged fine. It is possible to upgrade at least some packages in that version via pip (my only dependencies are NumPy and matplotlib, both of which can be updated this way). I do not use conda's updating system because that brings the same issues as using the latest condas and trying to make a build. If someone still needs this but the above solution does not work, is the error the same as those at the top of this issue or is it different?

**Comment by DiogoFerrari** [comment]
I had the same problem, and the solution [here](https://github.com/viotti/pyinstaller/commit/597af5cb01c064620b53ea1ee537e30f56fa481d) works partially. Even after using that workaround to use it on mac, the pyinstaller still doesn't copy the FILES tcl and tk which are located, in Sierra, on /System/Library/Frameworks/Tk.framework/Versions/8.5/Tk /System/Library/Frameworks/Tcl.framework/Versions/8.5/Tcl
respectively (for version 8.5 of tcl and tk). It is necessary to move them manually to the folder of the app created by pyinstaller. It means that the option --one-file can't be used.

**Comment by Ps2Fino** [comment]
Are there plans to fix this?
Is there at least a temporary fix to enable the --one-file option to be used?
My use case is for deploying applications to non-developers so I'm looking to avoid terminals and asking them to enter commands as much as possible

**Comment by LevN0** [comment]
The solution I described above works with --one-file. However it does mean using an outdated Python (or possibly not if other Anaconda issues with PyInstaller have since been fixed or have a workaround), although you should be able to upgrade at least some of the packages you need.

**Comment by Hyped-247** [comment]
@DiogoFerrari,  once the PyInstaller files are added to the working directory, then what is the terminal command that will create a TK app? Is it pyInstaller --onefile "filename"? If not, then what is it?

**Comment by DiogoFerrari** [comment]
@Hyped-247, if you add the files manually as I mentioned, then it means you have already ran the pyinstaller command without the --onefile. That is, you generate the app without --onefile, and then you copy the tk/tcl files to the folder of the app generated by the pyinstaller. I could not use the --onefile command in any way that worked.

**Comment by htgoebel** [comment]
Anybody to provide a pull-request?

**Comment by Hyped-247** [comment]
@DiogoFerrari , I am not very clear on the meaning of: "if you add the files manually as I mentioned, then it means you have already ran the pyinstaller command without the --onefile." Does it mean that once I add the pyinstaller files to my working directory, then it just creates the app without running: "--onefile"? If that is the correct meaning, then where is the "myapp.exe" file located?

**Comment by krid-sdnera** [comment]
I was able to get `--onefile` to work with a TK app.

Following @titan-ae 's comment https://github.com/pyinstaller/pyinstaller/issues/1350#issuecomment-191067609
and the subsequent link to a commit viotti/pyinstaller@597af5c with some further modifications.

I probably didn't go about it in the right way, however this is what i found.
I found that the as has been mentioned, the TK Framework is not being included in the `--onefile` bundle due to pyinstaller modifying `@loader_path/Tcl`.
This is because as @viotti describes the `tk` and `tcl` executables have the same name as the TK and TCL library folders.
Further to this i found that even with viotti's fix, the executables still weren't being copied. Hence the need for the new function to collect the execs.

For reference `hook_api.add_datas()` is simmilar to `--add-data`

I modified these two file:
`/pyinstaller/blob/develop/PyInstaller/hooks/hook-_tkinter.py`

Add this function somewhere:

    def _collect_tcl_tk_exe(hook_api):
        tcl_root, tk_root = _find_tcl_tk(hook_api)
      
        return (('Tcl',os.path.join(tcl_root,'..','..','Tcl'),'DATA'), ('Tk',os.path.join(tk_root,'..','..','Tk'),'DATA'))

Then change line 245 from `hook_api.add_datas(_collect_tcl_tk_files(hook_api))`

        datas  = _collect_tcl_tk_files(hook_api)
        if is_darwin:
            datas += _collect_tcl_tk_exe(hook_api)

        hook_api.add_datas(datas)

And then in
lines 222 and 224 of `/pyinstaller/blob/develop/PyInstaller/hooks/hook-_tkinter.py`
lines 24 and 25 of `/pyinstaller/blob/develop/PyInstaller/loader/rthooks/pyi_rth__tkinter.py`

change `tcl` to `tclResources`
change `tk` to `tkResources`



Note: I used this snippet to debug and list the files avaliable to the build app:

	import os,sys
	try:
		for root, subdirs, files in os.walk(sys._MEIPASS): 
			'''sys._MEIPASS is the location pyinstaller expands its temporary files into for you to access'''
			for file_ in files:
				print (root,file_)	
	except AttributeError:
		pass

I made this a couple of months ago and cant remember all the details
I also haven't finalized the project which required this fix so there is limited testing of this however it did work on my machine.

But I hope this helps others find a better, more robust fix.


Just quickly with @DiogoFerrari and @Hyped-247 
There are two way pyinstaller build from your source file, into "onefile" or into a folder.
Manually moving the TK and TCL folders, as DiogoFerrari is doing, is only an option when building into a folder and once there does not allow for building of a "onefile" app. Sorry.
Running pyinstaller again with `--onefile` will just build the whole app again. not what you were wanting.

**Comment by Hyped-247** [comment]
@krid-sdnera, thank you for the valuable feedback. Can you please provide a pull-request with all the changes that you have made?

**Comment by HarmPrinsDEX** [comment]
Hi, I had the same problem. I installed Python with Homebrew and after that installed pyinstaller, and it works like a dream. Thank you Homebrew!

**Comment by johnmlhll** [comment]
I have the same problem and posted in detail on it on stackoverflow at [https://stackoverflow.com/questions/48712076/import-error-tcl-tk-modules-not-found-pyinstaller-on-mac-not-working](https://stackoverflow.com/questions/48712076/import-error-tcl-tk-modules-not-found-pyinstaller-on-mac-not-working) 

Any thoughts on why even `.spec` manual imports in the `Analysis()` with `binaries=['/System/Library/Frameworks/Tcl.framework/Tcl'],` does not work either? I tried just Tcl, then added the Tk binary.. no change... ?? Any thoughts on same?

**Comment by htgoebel** [comment]
There are two unfinished pull-requests for this issue: #2969 and #3132. Both make it work on OS X but break all other platform. Someone of you needs to step forward to finish these pull-requests.

**Comment by HarmPrinsDEX** [comment]
Hi, I had the same problem. I installed Python with Homebrew and after that installed pyinstaller with pip3, and it works like a dream.

**Comment by djhyman** [comment]
That worked for me after **several** hours of trying. After installing Python3 via Homebrew and then installing pyinstaller via pip3, `pyinstaller myscript.py` would only work with Python2.7. I finally got it working with Python3 by using `python3 -m  PyInstaller myscript.py `

**Comment by saltycraig** [comment]
The following .spec file will get you a "folder" build (will try one-file build soon) working with Tcl/Tk 8.6 on Mac OS X 10.13.4 and Python 3.6.5. Here is my spec file contents, which I named 'osx-folder-build.spec'.

```
# -*- mode: python -*-

block_cipher = None

a = Analysis(['main.py'],
             pathex=['/Users/craig/projects/parou/src'],
             binaries=[('/System/Library/Frameworks/Tk.framework/Tk', 'tk'),
		       ('/System/Library/Frameworks/Tcl.framework/Tcl', 'tcl')],
             datas=[('./assets', 'assets'),
		    ('../README.md', 'docs'),
		    ('../LICENSE', 'docs'),
		    ('../CREDITS', 'docs')
	     ],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=True,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='parou',
          debug=False,
          strip=False,
          upx=True,
          console=False )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               name='parou')
```

Then run:

```$ pyinstaller osx-folder-build.spec```

I first did as others has done and moved Tcl & Tk binaries by hand (by writing a 'osx-build' Makefile target, actually), but adding to the binaries list worked for me.

**Comment by tanyanghan** [comment]
I had this exact problem with Mac OS X High Sierra 10.13.5 with `Python 2.7.15 (v2.7.15:ca079a3ea3, Apr 29 2018, 20:59:26)`. All you need to do is add the add-binary option in the pyinstaller command line:

```
pyinstaller --onefile \
            --add-binary='/System/Library/Frameworks/Tk.framework/Tk':'tk' \
            --add-binary='/System/Library/Frameworks/Tcl.framework/Tcl':'tcl' \
            some_app.py ```

**Comment by htgoebel** [comment]
Anybody to pick up and finish #2969?

**Comment by lalabuy948** [comment]
I have same issue on macOS High Sierra. `Python3.6.5`

I tried to use `--add-binary` flag with path to Tcl and Tk, still doesn't work for me.

```
ImportError: dlopen(/var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/_tkinter.cpython-36m-darwin.so, 2): Library not loaded: @loader_path/Tcl
  Referenced from: /var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/_tkinter.cpython-36m-darwin.so
  Reason: no suitable image found.  Did find:
	/var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/Tcl: not a file
	/var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/Tcl: stat() failed with errno=22
	/private/var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/Tcl: not a file
	/private/var/folders/qy/9by3pxwd1077w0gffyxn2lhm0000gn/T/_MEIRFcrzs/Tcl: stat() failed with errno=22
[15532] Failed to execute script app
```

**Comment by wsk170** [comment]
@lalabuy948 I use macOS High Sierra, Python3.6.6. and same issue.
here is my .spec file about add binaries. and it worked. you can try it. ignore bulid warning.
```
binaries=[
     ('/Library/Frameworks/Python.framework/Versions/3.6/lib/libtk8.6.dylib', 'tk'),
     ('/Library/Frameworks/Python.framework/Versions/3.6/lib/libtcl8.6.dylib', 'tcl')],
```

**Comment by lalabuy948** [comment]
@wsk170 Thank you for your comment. But i have absolutely same path to that binaries. And it doesnt work for me.

**Comment by ivanovaos** [comment]
Same here, providing binaries with `--add-binary` doesn't work. Python 3.6.3, mac os x 10.13.6

**Comment by v6** [comment]
// , I'm getting the same issue on Python 3.7.0.

**Comment by v6** [comment]
// , Here's an example of one of the errors: 
```
~ $/Users/v666/proj/docxtpl/dist/tktest/tktest ; exit;
Traceback (most recent call last):
  File "site-packages/PyInstaller/loader/rthooks/pyi_rth__tkinter.py", line 28, in <module>
FileNotFoundError: Tcl data directory "/Users/v666/proj/docxtpl/dist/tktest/tcl" not found.
[73791] Failed to execute script pyi_rth__tkinter
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

**Comment by 3dp** [comment]
Seeing this:
```
Traceback (most recent call last):
  File "PyInstaller/loader/rthooks/pyi_rth__tkinter.py", line 28, in <module>
FileNotFoundError: Tcl data directory "/var/folders/3p/7y007rf52kxcpg1t6rnky90wfrzr00/T/_MEIv2h6BH/tcl" not found.
[45261] Failed to execute script pyi_rth__tkinter
```
Been trying the various workarounds proposed with no success.  Only on High Sierra (10.13.6).  Windows 10 is fine.

**Comment by tazzben** [comment]
> Seeing this:
> 
> ```
> Traceback (most recent call last):
>   File "PyInstaller/loader/rthooks/pyi_rth__tkinter.py", line 28, in <module>
> FileNotFoundError: Tcl data directory "/var/folders/3p/7y007rf52kxcpg1t6rnky90wfrzr00/T/_MEIv2h6BH/tcl" not found.
> [45261] Failed to execute script pyi_rth__tkinter
> ```
> Been trying the various workarounds proposed with no success. Only on High Sierra (10.13.6). Windows 10 is fine.

Build the directory version which will create an app bundle that you can modify.  Within the bundle add empty tcl and tk directories.  Then copy over the other tcl/tk directories and files (including the libs and .sh).

**Comment by Hyped-247** [comment]
After a lot of failure, I came to learn that Pyinstaller is the worst option for mac. So, if you're a mac user, then I would recommend py2app. See [this](http://takluyver.github.io/posts/so-you-want-to-write-a-desktop-app-in-python.html) url since it recommends a specific Python library based on your operating system.

**Comment by melvyn2** [comment]
@Hyped-247 I don't find that to be true. I use PyInstaller fine for mac, it just seems to be that Tkinter isn't working too well here. Pyinstaller does work fine for many other libraries on mac.

**Comment by LevN0** [comment]
There are at least 3 different solutions to the issue with-in this thread. All of them work. None of them are extremely complicated. 

The easiest solution that I have personally used and thus know works or sure is [this one](https://github.com/krid-sdnera/pyinstaller/commit/9b118288d52ea7207692937a0341a3b9b3199fa3).

**Comment by Hyped-247** [comment]
@melvyn2 my experience with Pyinstaller has been different. Maybe there is something wrong with Tk as you wrote. And yes, Pyinstaller did work for me when I wanted to turn my python code into a terminal non-GUI program. However, when it comes to turning specifically your TK program into an executable in mac, I see that py2app is a much better option than Pyinstaller.

**Comment by 3dp** [comment]
@LevN0  - Thanks.  I'll try this.

**Comment by AmbroiseGrau** [comment]
> Thanks. I'll try this.

@3dp Did it work for you in the end? I found an other way to go around it if needed.

**Comment by Nutshell1** [comment]
This might be a simpler solution for all and seems to work like a charm for me. No homebrew or anaconda or copying files into folder or modifying any files needed. One file shareable app created. Works with both regular tkinter based gui or pysimplegui.

**Pyinstaller instructions Mac**

1) Navigate to directory where the script/s is/are saved using terminal

2) Copy and paste the following
pyinstaller --onefile --add-binary='/System/Library/Frameworks/Tk.framework/Tk':'tk' --add-binary='/System/Library/Frameworks/Tcl.framework/Tcl':'tcl' xyz.py

3) Remember to change the xyz and hit return

(Mac OS 10.11.6, python 3.7.1, pyinstaller 3.4)

**Comment by MarcelFerrari** [comment]
I managed to get PyInstaller working for OS X APP bundles. I wrote a tutorial on how fixed the problem, if you are interested you can find it here: https://marcelferrari.com/Bundle-PySimpleGui-scripts-as-OSX-APP-files-with-PyInstaller/

**Comment by haarc** [comment]
I ran into the same issue on MacOS 10.13.6. I use pyenv with virtualenvwrapper and created a new virtual environment just to use pyinstaller (and I also tried to freeze my Python app with fbs) with Python 3.6.0 (with --enable-framework option), PyInstaller 3.4 and PyQt5 5.9.2 which are recommended and have been reported to work. The result I get is still an app that does not open an actual window and I also get the debug output 'WARNING: failed to copy flags of /System/Library/Frameworks/Tk.framework/Versions/8.5/...'

**Comment by v6** [comment]
Hey, there's always Nim or GoLang if you want cross-platform compatibility and compiled binaries. Nim is even a scripting language. Not that this helps much for pre-existing projects, I guess.

**Comment by isolet** [comment]
None of the above solutions worked for me with mac 10.15.3 (19D76) and Python 3.7.6. Does anyone has any suggestion? Thanks!

**Comment by MarcelFerrari** [comment]
> None of the above solutions worked for me with mac 10.15.3 (19D76) and Python 3.7.6. Does anyone has any suggestion? Thanks!

I was not able to get PyInstaller working, but if you are not absolutely forced to use that you can try Platypus which is quite similar.

As I mentioned a few comments up, I’ve written this tutorial if you are interested: https://marcelferrari.com/Package-PySimpleGUI-scripts-as-Os-X-app-bundles-with-Platypus/

**Comment by tanyanghan** [comment]
If anyone isn't actually using tcl, tkinter, then you can add the `--exclude-module` argument to exclude the lot. 

```
pyinstaller --onefile \
            --exclude-module='FixTk' \
            --exclude-module='tcl' \
            --exclude-module='tk' \
            --exclude-module='_tkinter' \
            --exclude-module='tkinter' \
            --exclude-module='Tkinter' \
            your_progarm.py
```

**Comment by vinaywadhwa** [comment]
Either solutions (copy pasting to build folder OR using 'add binaries' option to build) do not work for me (OSX 10.15.5 , Python 3.6, PyInstaller 3.6)  .. any help would be appreciated!

**Comment by rainydew** [comment]
> This might be a simpler solution for all and seems to work like a charm for me. No homebrew or anaconda or copying files into folder or modifying any files needed. One file shareable app created. Works with both regular tkinter based gui or pysimplegui.
> 
> **Pyinstaller instructions Mac**
> 
> 1. Navigate to directory where the script/s is/are saved using terminal
> 2. Copy and paste the following
>    pyinstaller --onefile --add-binary='/System/Library/Frameworks/Tk.framework/Tk':'tk' --add-binary='/System/Library/Frameworks/Tcl.framework/Tcl':'tcl' xyz.py
> 3. Remember to change the xyz and hit return
> 
> (Mac OS 10.11.6, python 3.7.1, pyinstaller 3.4)

thank you, this worked for me

**Comment by rainydew** [comment]
> Either solutions (copy pasting binaries manually in to dist folder OR using 'add binaries' option to build) do not work for me (OSX 10.15.5 , Python 3.6, PyInstaller 3.6) .. any help would be appreciated!

did you run it as root?

**Comment by nathanielkam** [comment]
@rainydew, @Nutshell1 's solution worked for me and yes use sudo to run it, 

I am on 
144 INFO: PyInstaller: 3.6
144 INFO: Python: 3.7.6
169 INFO: Platform: Darwin-19.5.0-x86_64-i386-64bit
OSX 10.15.5

> This might be a simpler solution for all and seems to work like a charm for me. No homebrew or anaconda or copying files into folder or modifying any files needed. One file shareable app created. Works with both regular tkinter based gui or pysimplegui.
> 
> **Pyinstaller instructions Mac**
> 
> 1. Navigate to directory where the script/s is/are saved using terminal
> 2. Copy and paste the following
>    sudo pyinstaller --onefile --add-binary='/System/Library/Frameworks/Tk.framework/Tk':'tk' --add-binary='/System/Library/Frameworks/Tcl.framework/Tcl':'tcl' xyz.py
> 3. Remember to change the xyz and hit return
> 
> (Mac OS 10.11.6, python 3.7.1, pyinstaller 3.4)

**Comment by rokm** [comment]
The issue with `tkinter` in python versions that use system Tcl/Tk framework* was that pyinstaller overwrote the library paths to system `Tcl` and `Tk` shared library with path relative to the executable directory. Therefore copying the `Tcl` and `Tk` shared library into executable directory did work as a work-around, but the proper fix is to prevent the library path for those two libraries from being overwritten.

This should be fixed by commit 99c6e8bd07cb14dbb3e307d36ade0c021147cf37, so I'm closing this issue now. If your problem persists, please open a new issue with updated details.

(*NOTE: recent 64-bit only builds from python.org use their own bundled Tcl/Tk framework; see their release notes for details).
